core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeSorterSpillReader.java

```diff
@@ -104,6 +124,10 @@ public void loadNext() throws IOException {
     if (taskContext != null) {
       taskContext.killTaskIfInterrupted();
     }
+    if (this.din == null) {
+      // Good time to init (if all files are opened, we can get Too Many files exception)
+      initStreams();
+    }
```

Comment:
Can this solve the too many file open issue? When we do merging the readers, it is possibly that all the readers in priority queue still have records and are asked for records (so their files open). You still can encounter too many file open issue.

Reply:
Good point. PR has been tried with queries involving window functions (e.g Q67) for which it worked fine. 

During spill merges (esp getSortedIterator), it is possible to encounter too many open files issue.

Reply:
I think you need to first describe more about how to fix this issue in the description.

Reply:
I agree with @viirya , we're using priority queue to do merge sort, this will turn out to be all the readers in the priority queue is opened, so still cannot solve this issue.

I think a valid fix is to control the number of concurrent merged files, like MR's `io.sort.factor`.

Also we still need to address similar issue in `ExternalSorter` and other places in Shuffle.

Reply:
IIUC, this PR does not reduce the number of total open files. Since this PR tries to open files when they are required, this PR may reduce possibility of occurring an error of too may open files.

As @viirya pointed out, it is necessary to provide a feature to control the number of opening files at one point (e.g. priority queue).

Reply:
The valid fix should be to import a new config to control the concurrent number of opened spill files, it also means you should use some data structure to keep and track the request of open spill files.